File: web/src/main/kotlin/com/github/mustafaozhan/bookbarter/web/ui/MainView.kt
/*
 * Copyright (c) 2020 Mustafa Ozhan. All rights reserved.
 */

package com.github.mustafaozhan.bookbarter.web.ui

import com.github.mustafaozhan.bookbarter.client.di.getForJs
import com.github.mustafaozhan.bookbarter.client.main.MainViewModel
import com.github.mustafaozhan.bookbarter.web.app.AppDependenciesContext
import com.github.mustafaozhan.bookbarter.web.app.kermit
import react.RProps
import react.child
import react.dom.h1
import react.functionalComponent

//private val mainViewModel: MainViewModel by lazy {
//    useContext(AppDependenciesContext).koin.getForJs(MainViewModel::class)
//}

val MainView = functionalComponent<RProps> {
    kermit.d { "MainView" }
    child(
        functionalComponent {
            h1 { +"CCC" }
        }
    )
}

File: ios/ios/UI/MainView.swift
//
//  MainView.swift
//  ios
//
//  Created by Mustafa Ozhan on 28/01/2021.
//  Copyright © 2021 orgName. All rights reserved.
//

import SwiftUI
import NavigationStack
import client

typealias MainObservable = ObservableSEED<MainViewModel, BaseState, MainEffect, MainEvent, MainData>

struct MainView: View {

    @StateObject var observable: MainObservable = koin.get()

    @Environment(\.scenePhase) var scenePhase

    var body: some View {

        NavigationStackView(
            transitionType: .default,
            easing: Animation.easeInOut(duration: 0.5)
        ) {
            if observable.viewModel.isFistRun() {
                SliderView()
            } else {
                CalculatorView()
            }
        }
        .onAppear {observable.startObserving()}
        .onReceive(observable.effect) { onEffect(effect: $0) }
        .onChange(of: scenePhase) { phase in
            switch phase {
            case .background:
                LoggerKt.kermit.d(withMessage: {"App is in background"})
            case .active:
                LoggerKt.kermit.d(withMessage: {"App is Active"})
                observable.event.onResume()
            case .inactive:
                LoggerKt.kermit.d(withMessage: {"App is Inactive"})
                observable.event.onPause()
            @unknown default:
                LoggerKt.kermit.d(withMessage: {"New App state not yet introduced"})
            }
        }
    }

    private func onEffect(effect: MainEffect) {
        LoggerKt.kermit.d(withMessage: {effect.description})
        switch effect {
        default:
            LoggerKt.kermit.d(withMessage: {"unknown effect"})
        }
    }
}
